### PR TITLE
add jsmn (Jasmine) JSON parser

### PR DESCRIPTION
**Pull request by peterbarker**

### CODE REVIEW

**Comment by khancyr** [comment]
what is the advantage against picojson we are already using ?

**Comment by peterbarker** [comment]
> jsmm have no dynamic allocation

Yep.  No standard library either.  Really wanted json parsing on sitl-on-hardware, but had to cut it out.

**Comment by peterbarker** [comment]
I've done some more playing with this and this PR now replaced picojson with jasmine.

Cost of just having the jsmn in-play is 1424 bytes (i.e. create the parser and call it on a 1-byte string).

To actually walk through the tokens afterwards for Callisto.json to set parameters (in the same way that picojson did) costs an additional 1600 bytes.

File: Tools/autotest/models/Callisto.json [hunk 1]
```diff
@@ -57,5 +57,9 @@
     "mdrag_coef" : 0.10,
 
     # number of motors
```
Comment (tridge) [inline Tools/autotest/models/Callisto.json:59]: please please add support for '#' for comments, I know it is non-standard, but it is extremely useful

File: libraries/SITL/SIM_XPlane.h [hunk 2]
```diff
@@ -21,7 +21,7 @@
 #include <AP_HAL/AP_HAL_Boards.h>
 
 #ifndef HAL_SIM_XPLANE_ENABLED
-#define HAL_SIM_XPLANE_ENABLED (CONFIG_HAL_BOARD == HAL_BOARD_SITL)
+#define HAL_SIM_XPLANE_ENABLED 0
```
Comment (tridge) [inline libraries/SITL/SIM_XPlane.h:24]: umm, removing xplane?!?!?

**Review by tridge — CHANGES_REQUESTED** [review]
we need comments, and need xplane working

File: Tools/autotest/models/Callisto.json (hunk 1 above)
Comment (peterbarker) [reply Tools/autotest/models/Callisto.json:59]: > please please add support for '#' for comments, I know it is non-standard, but it is extremely useful

I swiped the existing "remove #" code - it is present, and does parse Callisto.json which has that sort of comment in it.

IOW - the support should already be there unless I'm missing something

File: libraries/SITL/SIM_XPlane.h (hunk 2 above)
Comment (peterbarker) [reply libraries/SITL/SIM_XPlane.h:24]: Just temporarily - this is a WIP and I wanted feedback before I did whatever needs doing to make xplane work again.

**Review by peterbarker — COMMENTED** [review]
> we need comments, and need xplane working

So you are happy with the way this is turning out and I should pursue it, then?  That's what this most recent push was in aid of - to make sure people didn't immediately throw up when they saw it.

I *believe* jasmine might be able to be run as a progressive parser, too - I haven't tried that, but it does have this:
```
  /* The string is not a full JSON packet, more bytes expected */
  JSMN_ERROR_PART = -3
```
... that would be rather awesome if it worked.

File: Tools/autotest/models/Callisto.json (hunk 1 above)
Comment (tridge) [reply Tools/autotest/models/Callisto.json:59]: ahh, sorry, saw the red in github markup and thought it was removing the comment line!

File: libraries/SITL/SIM_Frame.cpp [hunk 3]
```diff
@@ -341,12 +350,69 @@ void Frame::load_frame_params(const char *model_json)
     if (fname == nullptr) {
         AP_HAL::panic("%s failed to load\n", model_json);
     }
-    picojson::value *obj = (picojson::value *)load_json(model_json);
-    if (obj == nullptr) {
-        AP_HAL::panic("%s failed to load\n", model_json);
+
+    jsmntok_t t[50];
```
Comment (tridge) [inline libraries/SITL/SIM_Frame.cpp:354]: it should be good to have a simple load_json() method which stops us having to repeat all this logic in multiple places. I presume a wrapper like that, like we had with load_json() would be doable?

**Review by tridge — CHANGES_REQUESTED** [review]

**Comment by peterbarker** [comment]
Works for making a CubeOrange-SimOnHardware behave like a Callisto:
image

I've run this through Valgrind and fixed the one false-positive.

I've made sure that XPlane dref extraction works:
image

... and of course, Callisto still loads in SITL.

**Comment by peterbarker** [comment]
Flash cost for CubeOrange-SimOnHW is 3kB (+ whatever JSON you want in ROMFS e.g. 1kB for Callisto.json)

**Comment by tridge** [comment]
failed on libraries/SITL/examples/XPlane/xplane_alia.json

File: libraries/AP_JSONParser/AP_JSONParser.cpp [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+    if (t[token_offset].type != JSMN_PRIMITIVE) {
+        return false;
+    }
+
+    char tmp_value[20];
+    if (!get_string_for_current_token(tmp_value, ARRAY_SIZE(tmp_value))) {
+        return false;
+    }
+
+    ret = atoi(tmp_value);
```
Comment (Ryanf55) [inline libraries/AP_JSONParser/AP_JSONParser.cpp:159]: Do we ignore the `0` return failure here and then return `true`?

File: libraries/AP_JSONParser/AP_JSONParser.h [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+// defines functions.  If they're static that's not a problem, but
+// they also aren't used in the .h file, so we need to declare them
+// static:
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wunused-function"
+#define JSMN_STATIC
+#include "jsmn.h"
+#pragma GCC diagnostic pop
+
+#define AP_JSONPARSER_NUM_TOKENS 180  // allocated in cpp file
```
Comment (Ryanf55) [inline libraries/AP_JSONParser/AP_JSONParser.h:38]: static constexpr?

File: libraries/AP_JSONParser/jsmn.h [hunk 6]
```diff
@@ -0,0 +1,475 @@
+// this is the "JSMN" ("Jasmine") JSON parser.
+// it was directly copied from the github page here: https://raw.githubusercontent.com/zserge/jsmn/master/jsmn.h
+// thanks @Serge
```
Comment (Ryanf55) [inline libraries/AP_JSONParser/jsmn.h:3]: Can you add a commit hash or something so we know which version you pulled?

File: libraries/SITL/SIM_Frame.cpp [hunk 7]
```diff
@@ -322,10 +324,15 @@ float Frame::get_air_density(float alt_amsl) const
...
 }
 
-#if USE_PICOJSON
+#if AP_JSONPARSER_ENABLED
 /*
   load frame specific parameters from a json file if available
  */
+
+#ifndef streq
+#define streq(a,b) !strcmp(a, b)
```
Comment (Ryanf55) [inline libraries/SITL/SIM_Frame.cpp:333]: Does a macro provide any advantage over a constexpr function?

File: libraries/SITL/SIM_Frame.cpp [hunk 8]
```diff
@@ -384,63 +399,98 @@ void Frame::load_frame_params(const char *model_json)
...
-    };
-    char label_name[20];
-    for (uint8_t i=0; i<ARRAY_SIZE(per_motor_vars); i++) {
-        for (uint8_t j=0; j<12; j++) {
-            sprintf(label_name, "motor%i_%s", j+1, per_motor_vars[i].label);
-            auto v = obj->get(label_name);
-            if (v.is<picojson::null>()) {
-                // use default value
+        // check static list of known parameter names first:
+        for (auto &var : vars) {
```
Comment (Ryanf55) [inline libraries/SITL/SIM_Frame.cpp:422]: You could add const to `var`

**Comment by peterbarker** [comment]
> failed on libraries/SITL/examples/XPlane/xplane_alia.json

Fixed - it was just the maximum number of tokens we support in a file.

File: libraries/SITL/SIM_Frame.cpp (hunk 7 above)
Comment (peterbarker) [reply libraries/SITL/SIM_Frame.cpp:333]: > Does a macro provide any advantage over a constexpr function?

Possibly?  This is the way streq is usually formulated (and done in other places in the AP code).  Straight macro sub seems reasonable for this particular function to me.

File: libraries/SITL/SIM_Frame.cpp (hunk 8 above)
Comment (peterbarker) [reply libraries/SITL/SIM_Frame.cpp]: Done.

File: libraries/AP_JSONParser/jsmn.h (hunk 6 above)
Comment (peterbarker) [reply libraries/AP_JSONParser/jsmn.h:3]: Done, thanks!

File: libraries/AP_JSONParser/AP_JSONParser.h (hunk 5 above)
Comment (peterbarker) [reply libraries/AP_JSONParser/AP_JSONParser.h:38]: Or, you know, just remove it as it is unused :-)

File: libraries/AP_JSONParser/AP_JSONParser.cpp (hunk 4 above)
Comment (peterbarker) [reply libraries/AP_JSONParser/AP_JSONParser.cpp:159]: We are?!

There are integer conversion functions that can return error here.  I don't think we use them ATM in AP and I'm not going to chase that :-)

File: libraries/SITL/SIM_XPlane.h (hunk 2 above)
Comment (peterbarker) [reply libraries/SITL/SIM_XPlane.h]: XPlane is back

File: libraries/AP_JSONParser/AP_JSONParser.cpp (hunk 4 above)
Comment (Ryanf55) [reply libraries/AP_JSONParser/AP_JSONParser.cpp:159]: Thanks, just checking it was intentional. FYI, looks like `strtol` is already used in the codebase.

**Comment by peterbarker** [comment]
@Ryanf55 I've changed this to use strtol etc

**Comment by peterbarker** [comment]
Fixed XPlane.

**Comment by peterbarker** [comment]
Going with tridge's picojson changes instead.

**Pull request closed without merging**